src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIMarsHandle.cs

```diff
@@ -0,0 +1,521 @@
+            BitConverter.GetBytes(_currentHeader.flags).CopyTo(headerBytes, 1);
+            BitConverter.GetBytes(_currentHeader.sessionId).CopyTo(headerBytes, 2);
+            BitConverter.GetBytes(_currentHeader.length).CopyTo(headerBytes, 4);
+            BitConverter.GetBytes(_currentHeader.sequenceNumber).CopyTo(headerBytes, 8);
+            BitConverter.GetBytes(_currentHeader.highwater).CopyTo(headerBytes, 12);
+        }
+
+        /// <summary>
+        /// Generate a packet with SMUX header
+        /// </summary>
+        /// <param name="packet">SNI packet</param>
+        /// <returns>Encapsulated SNI packet</returns>
+        private SNIPacket GetSMUXEncapsulatedPacket(SNIPacket packet)
+        {
+            uint xSequenceNumber = _sequenceNumber;
+            byte[] headerBytes = null;
+            GetSMUXHeaderBytes(packet.Length, (byte)SNISMUXFlags.SMUX_DATA, ref headerBytes);
+
+            SNIPacket smuxPacket = new SNIPacket(null);
+            smuxPacket.Description = String.Format("({0}) SMUX packet {1}", packet.Description == null ? "" : packet.Description, xSequenceNumber);
+            smuxPacket.Allocate(16 + packet.Length);
+            smuxPacket.AppendData(headerBytes, 16);
+            smuxPacket.AppendPacket(packet);
+
+            return smuxPacket;
+        }
+
+        /// Send a packet synchronously
+        /// </summary>
+        /// <param name="packet">SNI packet</param>
+        /// <returns>SNI error code</returns>
+        public override uint Send(SNIPacket packet)
+        {
+            while (true)
+            {
+                lock (this)
+                {
+                    if (_sequenceNumber < _sendHighwater)
+                    {
+                        break;
+                    }
+                }
+
+                _ackEvent.WaitOne();
+
+                lock (this)
+                {
+                    _ackEvent.Reset();
+                }
+            }
+
+            return _connection.Send(GetSMUXEncapsulatedPacket(packet));
+        }
+
+        /// <summary>
+        /// Send packet asynchronously
+        /// </summary>
+        /// <param name="packet">SNI packet</param>
+        /// <param name="callback">Completion callback</param>
+        /// <returns>SNI error code</returns>
+        private uint InternalSendAsync(SNIPacket packet, SNIAsyncCallback callback)
+        {
+            SNIPacket encapsulatedPacket = null;
+
+            lock (this)
+            {
+                if (_sequenceNumber >= _sendHighwater)
+                {
+                    return TdsEnums.SNI_QUEUE_FULL;
+                }
+
+                encapsulatedPacket = GetSMUXEncapsulatedPacket(packet);
+
+                if (callback != null)
+                {
+                    encapsulatedPacket.SetCompletionCallback(callback);
+                }
+                else
+                {
+                    encapsulatedPacket.SetCompletionCallback(HandleSendComplete);
```

Comment:
I assume InternalSendAsync is called relatively frequently?  Each call that results in this SetCompletionCallback call is going to result in allocating a delegate for HandleSendComplete.